src/test/regress/spec/isolation_metadata_sync_vs_all.spec

```diff
@@ -0,0 +1,197 @@
+}
+
+step "s2-attach-partition"
+{
+	ALTER TABLE dist_partitioned_table ATTACH PARTITION dist_partitioned_table_p1 FOR VALUES FROM (1) TO (9);
+}
+
+step "s2-detach-partition"
+{
+	ALTER TABLE dist_partitioned_table DETACH PARTITION dist_partitioned_table_p1;
+}
+
+step "s2-create-partition-of"
+{
+	CREATE TABLE dist_partitioned_table_p2 PARTITION OF dist_partitioned_table FOR VALUES FROM (10) TO (20);
+}
+
+step "s2-create-type"
+{
+	CREATE TYPE my_type AS (a int, b int);
+}
+
+step "s2-drop-type"
+{
+	DROP TYPE my_type;
+}
+
+step "s2-create-dist-func"
+{
+	CREATE FUNCTION squares(int) RETURNS SETOF RECORD
+    AS $$ SELECT i, i * i FROM generate_series(1, $1) i $$
+    LANGUAGE SQL;
+
+	SELECT create_distributed_function('squares(int)');
+}
+
+step "s2-drop-dist-func"
+{
+	DROP FUNCTION squares(int);
+}
+
+session "s3"
+
+step "s3-compare-snapshot"
+{
+	SELECT bool_and(result::text[] @> activate_node_snapshot() AND result::text[] <@ activate_node_snapshot()) AS same_snapshot_across_cluster
+	FROM run_command_on_workers('SELECT activate_node_snapshot()');
+}
+
+step "s3-debug"
+{
+	SELECT unnest(activate_node_snapshot());
+
+	SELECT unnest(result::text[])
+	FROM run_command_on_workers('SELECT activate_node_snapshot()');
+}
+
+// the following operations get blocked when a concurrent metadata sync is in progress
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-alter-table" "s1-rollback" "s2-rollback" "s3-compare-snapshot"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-drop-table" "s1-rollback" "s2-rollback" "s3-compare-snapshot"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-dist-table" "s1-rollback" "s2-rollback" "s3-compare-snapshot"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-ref-table" "s1-rollback" "s2-rollback" "s3-compare-snapshot"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-attach-partition" "s1-rollback" "s2-rollback" "s3-compare-snapshot"
+permutation "s2-attach-partition" "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-detach-partition" "s1-rollback" "s2-rollback" "s3-compare-snapshot"
+permutation "s2-attach-partition" "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-partition-of" "s1-rollback" "s2-rollback" "s3-compare-snapshot"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-add-fk" "s1-commit" "s2-commit" "s3-compare-snapshot"
+permutation "s2-add-fk" "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-drop-fk" "s1-commit" "s2-commit" "s3-compare-snapshot"
+
+// the following operations do not get blocked
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-type" "s1-commit" "s2-commit" "s3-compare-snapshot" "s2-drop-type"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-dist-func" "s1-commit" "s2-commit" "s3-compare-snapshot" "s2-drop-dist-func"
+
+// the following operation creates a distributed deadlock and gets cancelled
```

Comment:
I'll check this

Reply:
I added one more deadlock case when dropping a distributed function concurrent with a `start_metadata_sync` operation.